### PR TITLE
feat: more basic function in sparrow py

### PR DESCRIPTION
**Pull request by jordanrfrazier**
adds:

`filter` (when)
`time_of`
`lag`
`with_key`
`lookup`
`if_` and `null_if`
`len_` or `length`

### CODE REVIEW

File: sparrow-py/pysrc/sparrow_py/_timestream.py [hunk 1]
```diff
@@ -552,6 +552,17 @@ def is_not_null(self) -> Timestream:
         """
         return Timestream._call("is_valid", self)
 
+    def filter(self, condition: Timestream) -> Timestream:
+        """
+        Create a Timestream containing only the points where `condition` is `true`.
```
Comment (bjchambers) [inline sparrow-py/pysrc/sparrow_py/_timestream.py:557]: Document Parameters too -- what is the `condition`

File: sparrow-py/pysrc/sparrow_py/_timestream.py [hunk 2]
```diff
@@ -576,6 +587,107 @@ def collect(
...
+        -------
+        Timestream
+            Timestream containing the length of `self`.
+        """
+        if self.data_type.equals(pa.string()):
+            return Timestream._call("len", self)
+        elif isinstance(self.data_type, pa.ListType):
+            return Timestream._call("list_len", self)
+        else:
+            raise NotImplementedError(f"length not implemented for {self.data_type}")
```
Comment (bjchambers) [inline sparrow-py/pysrc/sparrow_py/_timestream.py:651]: I think this should be `TypeError` -- not supported for that type

File: sparrow-py/pytests/filter_test.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import pytest
+import sparrow_py as kt
+
+
+@pytest.fixture(scope="module")
```
Comment (bjchambers) [inline sparrow-py/pytests/filter_test.py:5]: Thoughts on grouping these?

1. Fewer separate test files / directories of golden tests
2. Ability to share fixtures between tests in a file

Perhaps:
* `time_test` for `time_of` (we'll add other time stuff)
* `collection_test` for `length`
* `logical_test` or `conditional_test` for `if` and `null_if`

**Review by bjchambers — APPROVED** [review]

File: sparrow-py/pysrc/sparrow_py/_timestream.py (hunk 1 above)
Comment (jordanrfrazier) [reply sparrow-py/pysrc/sparrow_py/_timestream.py:557]: Done

**Pull request merged**